linkerd/app/core/src/serve.rs

```diff
@@ -1,66 +1,63 @@
+    listen: impl Stream<Item = std::io::Result<(Addrs, I)>>,
+    mut make_accept: M,
     shutdown: impl Future,
 ) -> Result<(), Error>
 where
-    C: HasSpan,
-    A: Accept<C> + Send + 'static,
+    I: Send + 'static,
+    M: tower::Service<Addrs, Response = A>,
+    M::Error: Into<Error>,
+    M::Future: Send + 'static,
+    A: tower::Service<I, Response = ()> + Send,
+    A::Error: Into<Error>,
     A::Future: Send + 'static,
-    A::Error: Send + 'static,
-    A::ConnectionError: Send + 'static,
-    A::ConnectionFuture: Send + 'static,
 {
-    // Initialize tracing & log errors on the accept stack.
-    let mut service = TraceAccept {
-        accept: AcceptError::new(accept),
-        span: Span::current(),
-    }
-    .into_service();
-
     let accept = async move {
         futures::pin_mut!(listen);
         loop {
             match listen.next().await {
                 None => return Ok(()),
                 Some(conn) => {
                     // If the listener returned an error, complete the task
-                    let conn = conn?;
+                    let (addrs, io) = conn?;
+
+                    // The local addr should be instrumented from the listener's context.
+                    let span = info_span!(
+                        "accept",
+                        peer.addr = %addrs.peer(),
+                    );
+                    let _enter = span.enter();
 
                     // Ready the service before dispatching the request to it.
                     //
                     // This allows the service to propagate errors and to exert backpressure on the
                     // listener. It also avoids a `Clone` requirement.
-                    let accept = service.ready_and().await?.call(conn);
+                    let accept = make_accept
+                        .ready_and()
+                        .err_into::<Error>()
+                        .await?
+                        .call(addrs);
 
                     // Dispatch all of the work for a given connection onto a connection-specific task.
                     tokio::spawn(
                         async move {
-                            match accept.await {
-                                Ok(serve) => match serve.await {
-                                    Ok(()) => {}
-                                    Err(e) => match e {},
+                            match accept.err_into::<Error>().await {
+                                Err(error) => error!(%error, "Failed to dispatch connection"),
+                                Ok(accept) => match accept.oneshot(io).err_into::<Error>().await {
+                                    Ok(()) => debug!("Connection closed"),
+                                    Err(error) => info!(%error, "Connection closed"),
                                 },
-                                Err(e) => match e {},
                             }
                         }
                         .in_current_span(),
```

Comment:
We shouldn't use `Span::enter` in cases where the `Entered` guard is held across an await point (see [here](https://docs.rs/tracing/0.1.18/tracing/span/struct.Span.html#in-asynchronous-code) for details).

Instead, we should write something like this:
```rust
                    // The local addr should be instrumented from the listener's context.
                    let span = info_span!(
                        "accept",
                        peer.addr = %addrs.peer(),
                    ); 

                    // Ready the service before dispatching the request to it.
                    //
                    // This allows the service to propagate errors and to exert backpressure on the
                    // listener. It also avoids a `Clone` requirement.
                    let accept = make_accept
                        .ready_and()
                        .err_into::<Error>()
                        .instrument(span.clone())
                        .await?
                        .call(addrs);


                    // Dispatch all of the work for a given connection onto a connection-specific task.
                    tokio::spawn(
                        async move {
                            match accept.err_into::<Error>().await {
                                Err(error) => error!(%error, "Failed to dispatch connection"),
                                Ok(accept) => match accept.oneshot(io).err_into::<Error>().await {
                                    Ok(()) => debug!("Connection closed"),
                                    Err(error) => info!(%error, "Connection closed"),
                                },
                            }
                        }
                        .instrument(span),
                    );
```
(not a suggested change because this contains deleted lines & github won't let me suggest changes).